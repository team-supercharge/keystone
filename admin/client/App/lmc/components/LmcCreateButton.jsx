import React, { Component } from 'react'
import PropTypes from 'prop-types'
import List from '../../../utils/List'
import CreateForm from '../../shared/CreateForm'
import { GlyphButton, ResponsiveText } from '../../elemental'

export default class LmcCreateButton extends Component {
    state = {
        currentList: null,
        isModalOpen: true,
        showCreateModal: false
    }

    onOpenCreateModal = () => {
        this.setState({ 
            showCreateModal: true,
            currentList: new List(Keystone.lists[this.props.listId]),
        })
    }

    onCloseCreateModal = () => {
        this.setState({ isModalOpen: false });
        setTimeout(() => {
            this.setState({ 
                showCreateModal: false, 
                isModalOpen: true 
            })
        })
    }

    onCreateItemComplete = (item) => {
        this.onCloseCreateModal()
        this.props.onCreate(item)
    }

    renderCreateForm = () => {
        const { prefillPath, prefillValue, title } = this.props
        const { showCreateModal, isModalOpen, currentList } = this.state

        return (
            showCreateModal
                ? <div className="lmc-create-form">
                    <CreateForm
                        isOpen={isModalOpen}
                        prefillPath={prefillPath}
                        prefillValue={prefillValue}
                        list={currentList}
                        onCancel={this.onCloseCreateModal}
                        formTitle={title}
                        onCreate={this.onCreateItemComplete}
                    />
                </div> : null
        )
    }

    render () {
        const { listId, title, style } = this.props
        const buttonText = `Add a ${listId}`
        return (
            <GlyphButton
                block
                color='success'
                glyph='plus'
                position='left'
                title={buttonText}
                onClick={this.onOpenCreateModal}
                style={style}
            >
                { this.renderCreateForm() }
                <ResponsiveText
                    visibleSM={buttonText}
                    visibleMD={buttonText}
                    visibleLG={buttonText}
                    />
            </GlyphButton>
        )
    }
}

LmcCreateButton.propTypes = {
    listId: PropTypes.string.isRequired,
    prefillPath: PropTypes.string,
    prefillValue: PropTypes.string,
    onCreate: PropTypes.func,
    styles: PropTypes.object,
    title: PropTypes.string.isRequired,
}