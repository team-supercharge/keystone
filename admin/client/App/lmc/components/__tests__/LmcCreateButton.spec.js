jest.mock('../../../shared/CreateForm')
jest.mock('../../../../utils/List')

import React from 'react'
import { shallow } from 'enzyme'
import LmcCreateButton from '../LmcCreateButton.jsx'

describe('LmcCreateButton', () => {
    let wrapper
    let listId
    let title
    let buttonText
    let styles
    const savedKeystone = global.Keystone
    const onCreate = jest.fn()

    beforeAll(() => {
        global.Keystone = { lists: { 'Resident': 'Resident' } }
    })

    beforeEach(() => {
        listId = 'TestList'
        title = 'TestTitle'
        buttonText = 'TestItem'
        styles = {
            button: {
                width: 200
            }
        }

        wrapper = shallow(
            <LmcCreateButton
                listId={listId}
                title={title}
                buttonText={buttonText}
                onCreate={onCreate}
                style={styles.button}
            />
        )
    })

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('renders the correct button', () => {
        const button = wrapper.find('GlyphButton')
        expect(button.props().title).toEqual('Add a TestItem')
        expect(button.props().style).toEqual(styles.button)
    })

    it('renders a CreateForm when the button is pressed', () => {
        const button = wrapper.find('GlyphButton')
        expect(wrapper.find('CreateForm').length).toEqual(0)

        button.simulate('click')
        expect(wrapper.find('CreateForm').length).toEqual(1)
    })

    afterEach(() => {
        onCreate.mockClear()
    })
    
    afterAll(() => {
        global.Keystone = savedKeystone
    })
})