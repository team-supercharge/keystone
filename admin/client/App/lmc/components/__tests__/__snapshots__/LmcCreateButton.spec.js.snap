// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcCreateButton renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <GlyphButton
    block={true}
    color="success"
    glyph="plus"
    glyphStyle={Object {}}
    onClick={[Function]}
    position="left"
    style={
      Object {
        "width": 200,
      }
    }
    title="Add a TestItem"
  >
    <ResponsiveText
      component="span"
      visibleLG="Add a TestItem"
      visibleMD="Add a TestItem"
      visibleSM="Add a TestItem"
    />
  </GlyphButton>,
  "nodes": Array [
    <GlyphButton
      block={true}
      color="success"
      glyph="plus"
      glyphStyle={Object {}}
      onClick={[Function]}
      position="left"
      style={
        Object {
          "width": 200,
        }
      }
      title="Add a TestItem"
    >
      <ResponsiveText
        component="span"
        visibleLG="Add a TestItem"
        visibleMD="Add a TestItem"
        visibleSM="Add a TestItem"
      />
    </GlyphButton>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LmcCreateButton
        buttonText="TestItem"
        listId="TestList"
        onCreate={[MockFunction]}
        style={
          Object {
            "width": 200,
          }
        }
        title="TestTitle"
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LmcCreateButton {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onCloseCreateModal": [Function],
        "onCreateItemComplete": [Function],
        "onOpenCreateModal": [Function],
        "props": Object {
          "buttonText": "TestItem",
          "listId": "TestList",
          "onCreate": [MockFunction],
          "style": Object {
            "width": 200,
          },
          "title": "TestTitle",
        },
        "refs": Object {},
        "renderCreateForm": [Function],
        "state": Object {
          "currentList": null,
          "isModalOpen": true,
          "showCreateModal": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <GlyphButton
          block={true}
          color="success"
          glyph="plus"
          glyphStyle={Object {}}
          onClick={[Function]}
          position="left"
          style={
            Object {
              "width": 200,
            }
          }
          title="Add a TestItem"
        >
          <ResponsiveText
            component="span"
            visibleLG="Add a TestItem"
            visibleMD="Add a TestItem"
            visibleSM="Add a TestItem"
          />
        </GlyphButton>,
        "_debugID": 2,
        "_renderedOutput": <GlyphButton
          block={true}
          color="success"
          glyph="plus"
          glyphStyle={Object {}}
          onClick={[Function]}
          position="left"
          style={
            Object {
              "width": 200,
            }
          }
          title="Add a TestItem"
        >
          <ResponsiveText
            component="span"
            visibleLG="Add a TestItem"
            visibleMD="Add a TestItem"
            visibleSM="Add a TestItem"
          />
        </GlyphButton>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcCreateButton
    buttonText="TestItem"
    listId="TestList"
    onCreate={[MockFunction]}
    style={
      Object {
        "width": 200,
      }
    }
    title="TestTitle"
  />,
}
`;
